### PR TITLE
fix(agent-guard): calibrate secret-detection patterns (close env/cloud-cred-dump gaps, fix dotenv-template & env-pipeline FPs)

### PR DESCRIPTION
**Pull request by JeongJaeSoon**
## Summary

Calibrates agent-guard's deterministic secret-detection denylist: closes a
cluster of environment / cloud-credential-dump **false negatives**, fixes two
**false positives** that blocked legitimate dev work, and hardens two backstops
that could silently no-op.

This is a deliberately coherent, reviewable slice driven by a multi-agent audit
against agent-guard's own design philosophy -- **thin but effective**: block
accidental or malicious secret exposure without over-blocking, stay
deterministic, fail closed. Scope is intentionally constrained: **no new
detection layer, no DLP/EDR expansion, no scanning of arbitrary file contents
referenced by shell.** Every change either tightens an existing denylist entry,
fixes an existing false positive, or makes an existing backstop actually run.

## Detection gaps closed

**`deny-bash-patterns.txt` / `deny-read-paths.txt`**

- **Process environment** -- reads of `/proc/<pid>/environ` (a direct bypass of
  the existing `env` / `printenv` blocks). The variable-PID form (`$$`,
  `$BASHPID`) the read-glob cannot express is handled in the bash list.
- **Cloud credential printers** -- `gcloud auth print-(access|identity)-token`,
  `az account get-access-token`, `aws configure export-credentials`,
  `doppler secrets|run`.
- **Kubernetes secret extraction** -- broadened to also catch `--output`, `-o=`,
  the no-space short form, the `secret/<name>` resource syntax, and
  `go-template`. Listing-only invocations (no output flag, or `-o name`) stay
  allowed by design.
- **Shell builtins that dump every variable** -- `export -p`, `declare -p`,
  `typeset -p`, and bare `set`. `set -e`, `set -o pipefail`, and
  `export FOO=bar` stay allowed.
- **Env exfil generalized** -- the env-pipe rule now matches any pipe or
  redirect rather than a fixed sink list, closing gaps like piping the
  environment into `gzip`/`gpg` or redirecting it to a file.
- **Additional secret file types** -- `*.p8`, `*.jks`, `*.keystore`, `.pgpass`,
  `.my.cnf`, `*.tfstate`, `*.tfvars`, `.vault-token`, `.dev.vars`.

## False positives fixed

- **`gitleaks.toml`** -- the allowlist had unanchored `x{8,}` and `0{12,}`
  regexes that suppressed *any* finding merely containing such a run, including
  real secrets. Anchored to placeholder-shaped strings only (verified against
  real gitleaks: the old config masked a real PAT, the new config flags it).
- **`.env` templates** -- `.env.example` / `.env.sample` / `.env.template` /
  `.env.dist` are committed by virtually every repo and were blocked by the
  `.env*` glob. Now allowlisted on both the Read and Bash paths via an exact
  basename match, while a bare `.env` (or any non-template `.env.*`) still
  blocks. The Bash path stays per-token, so `cat .env.example .env` still blocks
  on the real `.env` argument.

## Robustness

- **NUL-safe untracked-file backstop** -- `scan_untracked_files` now uses
  `git -c core.quotePath=false ls-files --others -z` with an `xargs -0` reader.
  The previous newline-delimited loop silently skipped files whose names git
  quotes (non-ASCII, backslash, quote, embedded newline) -- a place a careless
  or adversarial agent could park a secret the backstop never scanned.
- **Wrapper/assignment-aware git gate** -- `is_git_commit_or_push` now strips
  leading `NAME=value` assignments and no-op wrappers
  (`env`/`command`/`sudo`/`nice`/`nohup`/`time`/`xargs`, including their
  options) before the `git` check, so a wrapped commit/push still reaches the
  hook-bypass gate and the staged scan.

## Functional Verification

All checks run locally in this worktree against the modified tree.

```
$ make test
...
passed: 217
failed: 0
```

`make test` uses a mock gitleaks that flags the literal `AGENT_GUARD_TEST_SECRET`
(deterministic, offline) plus a real-gitleaks regression block. 13 new tests
were added, each shipping a **block** case and a **benign-passthrough** case:

- env-wrapped / assignment-prefixed / `env -u`-wrapped `git commit` against a
  staged secret all trigger the staged scan (exit 2); `env git status` stays a
  no-op (exit 0).
- kubectl variants `-o=yaml`, `--output yaml`, `secret/<name>`, `go-template`
  block; `get secrets` and `-o name` stay allowed.
- `env > file` and `env | gzip` block; a genuinely wrapped command pipeline
  (`env FOO=bar cmd | cat`) stays allowed.
- `cat .env.local` blocks (not a template suffix).
- **real gitleaks** still flags a `ghp_...` PAT that contains a 12-x run,
  confirming the anchored allowlist no longer masks real secrets.

```
$ make smoke-test
agent-guard: dependencies ok
agent-guard: gitleaks 8.30.0 (>= 8.30 recommended)
agent-guard: smoke ok: scan-path allows a clean directory
agent-guard: smoke ok: scan-path blocks a private-key fixture
agent-guard: smoke ok: hook-pre-tool blocks .env reads
agent-guard: smoke ok: pii-filter runs with the regex provider
agent-guard: smoke ok: hook-pre-tool blocks secret-like writes
agent-guard: smoke ok: native pre-commit hook blocks staged fixture
agent-guard: smoke-test ok
```

No VERSION bump here (release tooling handles that); configs unchanged for the
hooks.json matcher-parity test, which still passes.

## Deferred to a follow-up PR

Reviewed and ranked lower (docs / robustness / test-only); folding them in would
reduce this PR's reviewability: README/SECURITY blind-spot docs, gating
`hook_post_tool` on `git status --porcelain`, `json_value` fail-closed on
string `tool_input`, `bash_matches_deny_path_mode` fail-closed on invalid ERE,
and extending the hooks.json parity test to `Stop`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<details>
<summary>code-flow usage JSON</summary>

```json
{
  "commits": [
    {
      "confidence": "best_effort",
      "model_totals": [
        {
          "cache_creation_ephemeral_1h_input_tokens": 560279,
          "cache_creation_ephemeral_5m_input_tokens": 2041,
          "cache_creation_input_tokens": 562320,
          "cache_read_input_tokens": 8430142,
          "input_tokens": 45280,
          "model": "claude-opus-4-8",
          "output_tokens": 198928,
          "total_context_tokens": 9236670,
          "turns": 101
        }
      ],
      "participants": [
        {
          "attribution": "exact",
          "cache_creation_ephemeral_1h_input_tokens": 560279,
          "cache_creation_ephemeral_5m_input_tokens": 2041,
          "cache_creation_input_tokens": 562320,
          "cache_read_input_tokens": 8430142,
          "input_tokens": 45280,
          "kind": "main",
          "model": "claude-opus-4-8",
          "name": "main",
          "output_tokens": 198928,
          "total_context_tokens": 9236670,
          "turns": 101
        }
      ],
      "recorded_at": "2026-06-11T00:42:31Z",
      "sha": "b65f1c1cd662cc67538f679f05cd3d4625d6c19d",
      "subject": "fix(agent-guard): calibrate secret-detection patterns",
      "totals": {
        "cache_creation_ephemeral_1h_input_tokens": 560279,
        "cache_creation_ephemeral_5m_input_tokens": 2041,
        "cache_creation_input_tokens": 562320,
        "cache_read_input_tokens": 8430142,
        "input_tokens": 45280,
        "output_tokens": 198928,
        "total_context_tokens": 9236670,
        "turns": 101
      }
    },
    {
      "confidence": "best_effort",
      "model_totals": [
        {
          "cache_creation_ephemeral_1h_input_tokens": 136030,
          "cache_creation_input_tokens": 136030,
          "cache_read_input_tokens": 2552557,
          "input_tokens": 1384,
          "model": "claude-opus-4-8",
          "output_tokens": 30515,
          "total_context_tokens": 2720486,
          "turns": 24
        }
      ],
      "participants": [
        {
          "attribution": "exact",
          "cache_creation_ephemeral_1h_input_tokens": 136030,
          "cache_creation_input_tokens": 136030,
          "cache_read_input_tokens": 2552557,
          "input_tokens": 1384,
          "kind": "main",
          "model": "claude-opus-4-8",
          "name": "main",
          "output_tokens": 30515,
          "total_context_tokens": 2720486,
          "turns": 24
        }
      ],
      "recorded_at": "2026-06-11T00:52:26Z",
      "sha": "70e8def5cc9c8d52ff2d084e945c93687d286de1",
      "subject": "fix(agent-guard): match kubectl output flag on either side of the resource",
      "totals": {
        "cache_creation_ephemeral_1h_input_tokens": 136030,
        "cache_creation_input_tokens": 136030,
        "cache_read_input_tokens": 2552557,
        "input_tokens": 1384,
        "output_tokens": 30515,
        "total_context_tokens": 2720486,
        "turns": 24
      }
    }
  ],
  "confidence": "best_effort",
  "metered_commits": 2,
  "pr": {
    "base": "main",
    "head": "fix/detection-calibration",
    "number": 59
  },
  "schema": "code-flow-usage/1",
  "scope": "current_branch_pr",
  "source": "claude_code_hooks",
  "total_commits": 2,
  "totals": {
    "cache_creation_ephemeral_1h_input_tokens": 696309,
    "cache_creation_ephemeral_5m_input_tokens": 2041,
    "cache_creation_input_tokens": 698350,
    "cache_read_input_tokens": 10982699,
    "input_tokens": 46664,
    "output_tokens": 229443,
    "total_context_tokens": 11957156,
    "turns": 125
  },
  "totals_by_model": [
    {
      "cache_creation_ephemeral_1h_input_tokens": 696309,
      "cache_creation_ephemeral_5m_input_tokens": 2041,
      "cache_creation_input_tokens": 698350,
      "cache_read_input_tokens": 10982699,
      "input_tokens": 46664,
      "model": "claude-opus-4-8",
      "output_tokens": 229443,
      "total_context_tokens": 11957156,
      "turns": 125
    }
  ],
  "updated_at": "2026-06-11T00:52:35Z"
}
```

</details>

## Release Notes

* **New Features**
  * Expanded secret detection for additional credential tools (AWS, Azure, Google Cloud, Doppler)
  * Enhanced detection of environment variable dumping and exfiltration attempts
  * Added Kubernetes secret extraction blocking

* **Bug Fixes**
  * Improved handling of special characters and Unicode in filenames
  * Better parsing of environment-variable and wrapper-prefixed commands
  * Refined placeholder exemption logic

* **Tests**
  * Expanded regression test coverage for edge cases and secret scanning robustness

### CODE REVIEW

File: plugins/agent-guard/config/deny-bash-patterns.txt [hunk 1]
```diff
@@ -1,14 +1,39 @@
...
+aws[[:space:]]+configure[[:space:]]+export-credentials
 gcloud[[:space:]]+secrets[[:space:]]+versions[[:space:]]+access
+gcloud[[:space:]]+auth[[:space:]]+print-(access|identity)-token
+az[[:space:]]+account[[:space:]]+get-access-token
+(^|[[:space:];|&])doppler[[:space:]]+(secrets|run)
+# Reading secret material out of kubectl requires an output flag; plain
+# `kubectl get secrets` (listing names only, or `-o name`) is intentionally left
+# allowed. Covers `secret/name` syntax, `-o`/`--output`, `-o=`/`-ojson` forms,
+# and `go-template` extraction.
+kubectl[[:space:]].*get[[:space:]]+secret(s)?(/[^[:space:]]+)?([[:space:]].*)?(-o|--output)[[:space:]=]*(yaml|json|jsonpath|go-template)
```
[the remark this answers is not in the document]
Comment (JeongJaeSoon) [reply plugins/agent-guard/config/deny-bash-patterns.txt]: Thanks for catching this — you're right. kubectl accepts the output flag on either side of the resource (`kubectl get [(-o|--output=)yaml|...] (TYPE...)`), so the original single-order rule blocked `secret/x -o yaml` but missed the reordered `-o yaml secret/x`, which dumps the full Secret all the same.

Fixed in commit 70e8def by covering both orders with two token-anchored patterns:

- order A: secret resource, then `-o`/`--output` `yaml|json|jsonpath|go-template`
- order B: the output flag, then the secret resource

Both anchor the secret resource on a token boundary (a leading space), so a name that merely contains "secret" (e.g. a configmap `app-secret`) doesn't false-positive, and listing-only forms (`get secrets`, `-o name`) stay allowed. One residual over-block is documented inline: a namespace literally named `secrets` combined with `-o yaml` — rare enough that I favored closing the dump.

Added regression tests for both reordered forms (blocked) plus the configmap/pods cases with structured output (allowed). `make test` → 221/221.

Written by claude code assistant
